mule-extensions-api/src/main/java/org/mule/runtime/extension/api/OnTerminateResult.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.extension.api;
+
+import org.mule.runtime.api.message.Error;
+import org.mule.runtime.extension.api.annotation.execution.OnTerminate;
+
+import java.util.function.Consumer;
+
+/**
+ * Result for the {@link OnTerminate} callbacks.
+ * This result provides an easy way and common way to discern the different kinds of states that a
+ *
+ * @since 1.0
+ */
+@FunctionalInterface
+public interface OnTerminateResult {
+
+  void execute(Runnable onSuccess, Consumer<Error> onParameterResolutionError, Consumer<Error> onResponseError);
```

Comment:
Agree with the 1st, 2nd and 4th point, but:
* Why it doesn't support all the possible applications?

About new `OnTerminate{SomeName}`:
* The `resultMessage` should not be exposed, doing this is the same as having `Message` as a parameter of an Operation. Be aware of what is in the payload or attributes doesn't give you any clear idea of how you source OnSuccess/OnError callbacks were configured.